File: python/src/blind150/stacks/ValidParenthesis.py
from src.utils.logging_util import get_logger

logger = get_logger(__name__)
class ValidParenthesis:
    def isValid(self, s: str) -> bool:
        logger.info("isValid method invoked with input: %s", s)
        stack = []
        _map = {"(": ")", "{": "}", "[": "]"}

        for c in s:
            logger.debug("Processing character: %s", c)
            if c in _map:
                if stack and stack[-1] != c:
                    stack.pop()
                else:
                    logger.warning("Mismatch found for opening bracket %s. Returning False.", c)
                    return False
            else:
                logger.debug("Pushing character %s to stack.", c)
                stack.append(c)

        is_valid = not stack
        logger.info("Validation complete. Result: %s", "Valid" if is_valid else "Invalid")
        return is_valid

# Driver Code
if __name__ == "__main__":
    validator = ValidParenthesis()

    test_inputs = [
        "({[]})",
        "([)]",
        "",
        "({[",
        "[]",
    ]

    for i, test_input in enumerate(test_inputs, 1):
        logger.info("Test Case %d: Starting validation for input: %s", i, test_input)
        result = validator.isValid(test_input)
        logger.info("Test Case %d Result: Input '%s' is %s", i, test_input, "Valid" if result else "Invalid")

File: java/src/main/java/dsa/blind150/stacks/ValidParenthesis.java
package dsa.blind150.stacks;

import dsa.utils.LoggerUtil;
import org.slf4j.Logger;

import java.util.HashMap;
import java.util.Map;
import java.util.Stack;

public class ValidParenthesis {
    private static final Logger logger = LoggerUtil.getLogger();

    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();

        Map<Character, Character> map = new HashMap<>();
        map.put('(', ')');
        map.put('[', ']');
        map.put('{', '}');

        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (map.containsKey(c)) {
                if(stack.isEmpty() || stack.peek() != map.get(c)) {
                    return false;
                }else{
                    stack.pop();
                }
            }else{
                stack.push(c);
            }
        }
        return stack.isEmpty();
    }

    public static void main(String[] args) {
        ValidParenthesis validator = new ValidParenthesis();

        String[] testInputs = {
                "({[]})",
                "([)]",
                "",
                "({[",
                "[]",
                "{[()]}"
        };

        for (int i = 0; i < testInputs.length; i++) {
            String testInput = testInputs[i];
            logger.info("Test Case {}: Starting validation for input: {}", i + 1, testInput);
            boolean result = validator.isValid(testInput);
            logger.info("Test Case {} Result: Input {} is {}", i + 1, testInput, result ? "Valid" : "Invalid");
        }
    }

}
